pkg/scheduler/scheduler.go

```
@@ -20,24 +22,58 @@ import (
 	"gopkg.in/mgo.v2"
 	"gopkg.in/mgo.v2/bson"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
 	"k8s.io/client-go/util/retry"
 	"k8s.io/klog/v2"
 )
 
 const (
-	databaseNameRunningJobs = "runnings"
-	reschedChannelSize      = 100
-	restartChannelSize      = 100
-	// Number of GPUs of this scheduler
-	// TODO: should be user specified or discover in runtime
-	gpus = 8
-
+	databaseNameRunningJobs     = "runnings"
+	reschedChannelSize          = 100
+	restartChannelSize          = 100
 	rateLimitTimeMetricsSeconds = 5
 	reschedRateLimitSeconds     = 30
 )
 
+func DiscoverGPUs() int {
+	var kubeconfig *string
+
+	if home := homedir.HomeDir(); home != "" {
+		kubeconfig = flag.String("kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) absolute path to the kubeconfig file")
+	} else {
+		kubeconfig = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	}
+	flag.Parse()
+
+	// use the current context in kubeconfig
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// create the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		panic(err.Error())
```

It's better to return an error rather than `panic` here.